Browser_IDE/editorMain.js

```diff
@@ -2,38 +2,350 @@
+                return;
+            }
+
+            // ignore rename if still unsaved
+            let oldExists = await storedProject.access((project) => project.exists(self.filename));
+            if (oldExists) {
+                await storedProject.access((project) => project.rename(self.filename, newFilename));
+                // don't really care if the execution environment fails - should the code be mirrored there in the first place?
+                try {
+                    executionEnviroment.rename(self.filename, newFilename);
+                } catch {};
+            }
+
+            self.filename = newFilename;
+            self.basename = basename;
+
+            self.label.innerText = self.basename;
+        }
+        catch(err) {
+            let errEv = new Event("filesystemError");
+            errEv.shortMessage = "Rename failed";
+            errEv.longMessage = "An error occured and " + self.filename + " could not be renamed to " + newFilename + ".\n\nReason:\n" + err;
+            window.dispatchEvent(errEv);
+            return;
+        }
+    }
+
+    close() {
+        this.tab.remove();
+        this.editorContainer.remove();
+
+        this.filename = null;
+        this.basename = null;
+        this.label = null;
+        this.tab = null;
+        this.editorContainer = null;
+        this.editor = null;
+    }
+
+}
+
+function getExtension(filename) {
+    return filename.slice(filename.lastIndexOf('.')+1);
+}
+
+async function getFilesByExtension(extensions) {
+    let files = await storedProject.access((project) => project.getFlatFileList());
+
+    return files.filter((filename) => extensions.indexOf(getExtension(filename)) != -1);
+}
+
+async function findAllSourceFiles() {
+    return await getFilesByExtension(activeLanguage.sourceExtensions);
+}
+
+async function findAllCompilableFiles() {
+    return await getFilesByExtension(activeLanguage.compilableExtensions);
+}
+
+let editors = [];
+
+function getCodeEditor(filename) {
+
+    for(let i = 0; i < editors.length; i ++) {
+        if (editors[i].filename == filename)
+            return editors[i];
+    }
+
+    return null;
+}
+
+function openCodeEditor(filename, setFocus=true, load=true) {
+    let existing = getCodeEditor(filename);
+    if (existing) {
+        if (setFocus)
+            SwitchToTab(existing);
+
+        return;
+    }
+
+    let codeView = new CodeViewer(filename);
+
+    if (load)
+        codeView.load();
+
+    editors.push(codeView);
+
+    if (setFocus) {
+        SwitchToTab(codeView);
+    }
+}
+
+async function updateNoEditorsMessage() {
+    document.getElementById("noEditorsMessage").style.opacity = (editors.length == 0 && !makingNewProject) ? 1 : 0;
+}
+
+async function openCodeEditors() {
+    let sourceFiles = await findAllSourceFiles();
+
+    for(let i = 0; i < sourceFiles.length; i ++) {
+        openCodeEditor(sourceFiles[i], false);
+    }
```

Comment:
I was wondering about this, but right now I think it's best to immediately show code upon page load - otherwise a first time user will need to figure out they can open them from the file tree. 
This might seem obvious, but people aren't always as perceptive as one would hope, so I want to keep first impressions clear. 
I'll definitely add a limit to the number it loads though (max 3 maybe?), as a stop-gap until we store in the projects which ones were last open.